### ISSUE TITLE
randomForestSRCSyn learners are broken

### ISSUE THREAD

**Issue by larskotthoff** [mlr-org/mlr#602]
Tests fail:

    1. Error: classif_randomForestSRCSyn -------------------------------------------
     dim(X) must have a positive length
     1: withCallingHandlers(eval(code, new_test_environment), error = capture_calls, message = function(c) invokeRestart("muffleMessage"))
    2: eval(code, new_test_environment)
    3: eval(expr, envir, enclos)
    4: apply(mod$predicted, MARGIN = 1L, which.max) at test_classif_randomForestSRCSyn.R:21
    5: stop("dim(X) must have a positive length")

etc.

randomForestSRC package got updated, not obvious to me what changed.

**Comment by berndbischl** [comment]
Like I said in the PR where we noticed the error first:

We predict on one observation, I guess. Now, in the new package version of rfSRCSyn, a the dimension gets dropped and we get a single vector instead of a df / matrix. Therefore our apply breaks.

It is annoying, but the fix is a simple "if" case.

**Comment by berndbischl** [comment]
I will do this now

**Comment by berndbischl** [comment]
I look at the code. They basically did a renaming of some slots.

BUT: The orig. code is really strange. We apparently cannot predict  predict on newdata, if newdata does not contain the target label col. So the person wrapping this, actually TRAINED the model during predict. 
I am very close to voting for learner removal as I currently cannot see how this can be properly integrated.

Can I get more opinions fast pls?

**Comment by berndbischl** [comment]
Here code to look at:

```
library(mlbench)
require("randomForestSRC")
load_all()

ring = data.frame(mlbench.ringnorm(25, 2))
ringSyn1 = rfsrcSyn(classes ~., data = ring)
nd = ring[, 1:2]
p = rfsrcSyn(object = ringSyn1, newdata = nd)$rfSynPred
print(str(p))
```

**Comment by berndbischl** [comment]
I will remove the learners to todofiles now, so travis runs again

**Comment by schiffner** [comment]
I only had a short look:
With the previous version 1.6.1 you actually *could* predict on newdata *without* the target column. (So there was really no need to train the model in predictLearner in the first place).
I guess it's time for a bug report if we want this solved properly.

**Comment by berndbischl** [comment]
@schiffner Would you do this and ask them about this?

**Comment by schiffner** [comment]
I feared that you would say this. :)
I can ask them.

**Comment by berndbischl** [comment]
I fixed more SRC problems.

A side issue is that we have hardcoded args again:
```
trainLearner.classif.randomForestSRC = function(.learner, .task, .subset, .weights = NULL,  ...) {
  f = getTaskFormula(.task)
  randomForestSRC::rfsrc(f, data = getTaskData(.task, .subset, recode.target = "drop.levels"), importance = "none", proximity = FALSE, forest = TRUE, ...)
}

```

I guess we should rather set this in parvals

**Comment by berndbischl** [comment]
@schiffner 
The most important problem is that the "surv" SRC forest is still broken. They NOW return only predictions for the train set? 
At least the dimensionality of the "predicted" slot seems to indicate that. 
Can you please set me in CC? It really looks like another bug....?

**Comment by zmjones** [comment]
Talked with @berndbischl some.

Although the default value of `outcome` is `c("train", "test")`, in `generic.predict.rfsrc` (which `predict.rfsrc` calls) `match.arg` is called on this and so `outcome = "train"` if the user doesn't set `outcome = "test"`. This argument controls whether the prediction at each terminal node is recomputed using the input to `newdata`. So when `outcome = "train"` and `newdata` is passed, you get the normal random forest type of prediction, which is the predictions made by the terminal nodes from the training data. With `outcome = "test"` you get something like RF KNN. That is why when `outcome = "test"` `newdata` must also contain y-values.

All this is a long way of saying that our code should work. Specifying `newdata` and leaving `outcome` unspecified (which opaquely defaults to "train") should mean that the `predicted` element of the return object has the same length as the number of rows in `newdata`. 

I also see nothing in the NEWS for randomForestSRC that indicates that their API has changed for univariate survival tasks.

**Comment by schiffner** [comment]
Looked into it some and agree.

So I will ask about these two problems:

1) For survival forests slot `predicted` seems to contain predictions on the training data (although the documentation says test data and `predict.rfsrc` works in this way for regression and classification).

```
library(randomForestSRC)

data(veteran, package = "randomForestSRC")
veteran.test = veteran[1:10,]

obj = rfsrc(Surv(time, status) ~ ., data = veteran, ntree = 100)
pred = predict(obj, newdata = veteran.test)

 # $ chf          : num [1:10, 1:97] 0.005 0 0.0225 0 0.00333 ...
 # $ chf.oob      : NULL
 # $ predicted    : num [1:137] 67.3 27.8 41.4 55.7 60.9 ...
 # $ predicted.oob: NULL
 # $ survival     : num [1:10, 1:97] 0.995 1 0.978 1 0.997 ...
 # $ survival.oob : NULL
```

2) With the synthetic forest it does not seem to be possible to make predictions on a test data set that does not contain a target column / y-values.

```
library(mlbench)
ring = data.frame(mlbench.ringnorm(25, 2))
ring.test = ring[1:10, 1:2]

obj = rfsrcSyn(classes ~., data = ring)
pred = rfsrcSyn(object = obj, newdata = ring.test)
# 	 making the test set synthetic features
# Error in `[.data.frame`(newdata, , yvar.names, drop = FALSE) : 
#   undefined columns selected
traceback()
# 5: stop("undefined columns selected")
# 4: `[.data.frame`(newdata, , yvar.names, drop = FALSE)
# 3: newdata[, yvar.names, drop = FALSE]
# 2: data.frame(newdata[, yvar.names, drop = FALSE], x.s = xtest.s)
# 1: rfsrcSyn(object = obj, newdata = ring.test)
```

Did I miss anything?

**Comment by berndbischl** [comment]
Yes, that is also what I got from my analysis.

**Comment by schiffner** [comment]
This seems to be fixed now in randomForestSRC 2.0.5.

**Comment by berndbischl** [comment]
PR was merged but there are still 2 open TODOs

1) check thread again, there was something about hardcoded params
2) check whether NEWS is OK

**Comment by schiffner** [comment]
Some initial feedback:

I've only looked at `classif.randomForestSRC` yet (which is not directly subject of this issue or the PR, but which I understand best). As far as I know yet (need to read the docs again when I'm more awake):

1) In my opinion not all of the hardcoded params need to be hardcoded (better put them in `par.vals` with a `note`).
* Hardcoding `forest = TRUE` is ok because otherwise no forest object is returned and no predictions can be made.
* `importance = FALSE` does not hurt because the docs say that " variable importance calculations can always be recovered later from the grow forest using the function vimp and/or predict", but I prefer putting the default in `par.vals` and let the user decide when to calculate the importance.
* Hardcoding `proximity = FALSE` is not ok, because there seems to be no convenience function to calculate it afterwards.

2) There are some learner params missing in the `par.set` and I suspect that there might be missing dependencies.

3) Learner property `"ordered"` is missing.

**Comment by berndbischl** [comment]
> Some initial feedback:
> 
> I've only looked at classif.randomForestSRC yet (which is not directly subject of this issue or the PR, but which I understand best). As far as I know yet (need to read the docs again when I'm more awake):
> 
> 1) In my opinion not all of the hardcoded params need to be hardcoded (better put them in par.vals with a note).
> 
> Hardcoding forest = TRUE is ok because otherwise no forest object is returned and no predictions can be made.
> importance = FALSE does not hurt because the docs say that " variable importance calculations can always be recovered later from the grow forest using the function vimp and/or predict", but I prefer putting the default in par.vals and let the user decide when to calculate the importance.
> Hardcoding proximity = FALSE is not ok, because there seems to be no convenience function to calculate it afterwards.

So we simply add both to par.vals? 

> 2) There are some learner params missing in the par.set and I suspect that there might be missing dependencies.

Can we fix this NOW or shall we open up a small issue for this?

> 3) Learner property "ordered" is missing

So simply add this?

**Comment by berndbischl** [comment]
@schiffner 

I really need to upload and check the package as a whole now. 
Would you take over this for 2.9, you seem to have already figured out what is wrong.
Simply adapt and do a PR.

**Comment by schiffner** [comment]
Ok, will do.
Sorry for not getting back to this sooner.

**Comment by schiffner** [comment]
The randomForestSRC learners are done. Will do the randomForestSRCSyn learners next.

**Comment by schiffner** [comment]
I fixed the randomForestSRCSyn learners, i.e., fixed the param set, hardcoded args and tried to get rid of the strange structure (training the model during prediction). It works for regression.
For classification with >2 classes I encountered some problems with the randomForestSRC package and emailed the maintainers.

**Comment by larskotthoff** [comment]
Great, thanks. Do you want to make a PR for the stuff so far or wait until it's fixed upstream?

**Comment by schiffner** [comment]
At the moment my PR wouldn't pass all tests and therefore I tend to wait a bit. The last time they responded real fast.

**Comment by schiffner** [comment]
A new version of randomForestSRC will be on CRAN in 3 weeks at the earliest.
In the meantime I can issue a PR for the regression case and also update the param.set for the classification learner.

**Comment by larskotthoff** [comment]
Yeah, uh, we'll DEFINITELY have a new mlr release before then (maybe), so I guess there's no rush as this would only make it into the release after the release that's happening real soon now (no really).

**Comment by schiffner** [comment]
When is real soon?

**Comment by larskotthoff** [comment]
I think only Bernd and God know, but I'm not sure about God.

**Comment by schiffner** [comment]
:laughing:

**Comment by larskotthoff** [comment]
And He moves in mysterious ways.

**Comment by schiffner** [comment]
> And He moves in mysterious ways.
Indeed, indeed ...

But in the meantime I managed to do the regression stuff.

**Comment by berndbischl** [comment]
@schiffner 
Can we close this? Or do you also need to do the changes to classif? Can we also do this soon, to be consistent? I guess this is basically the same code change?

**Comment by schiffner** [comment]
We need the same changes for the classif case.
Actually, I have already done this, but the tests fail because of a problem with randomForestSRC.
A new version of randomForestSRC will be on CRAN in about 3 weeks.

What do you prefer?
* wait 3 weeks
* I could update the par.set for classif now and do the rest later
* I could issue a PR for all my changes and disable the tests for the multiclass case

**Comment by berndbischl** [comment]
> I could issue a PR for all my changes and disable the tests for the multiclass case

So this would imply we are done here? And add code can already be merged? Then I would prefer this.
Simply open a mini issue so we enable the test in 3 weeks.

**Comment by schiffner** [comment]
ok.

**Comment by schiffner** [comment]
@berndbischl : 
The new version of randomForestSRC is not yet on CRAN.
So I just rebased PR #832 (where some tests are disabled) if you want to pull it in for the release.

**Comment by berndbischl** [comment]
Julia, I would like to merge this in.

a) is this in your opionion ok for you?
b) could you please close this then, when I have I merged, and simply open a clean mini-issue for remaining stuff (like that we have to turn the tests back on, etc)

?

**Comment by schiffner** [comment]
Re a) 
Pro: Consistency with regression.

Con: The multiclass random forest does not work properly. I disabled the tests to  make sure that we on't break on CRAN. The only thing that could go wrong is that somebody tries to fit a multiclass forest and complains... For me that's ok because the new randomForestSRC release should be on CRAN in a few days.

Re b) The issue for the tests is #831 .

**Comment by schiffner** [comment]
A new version of randomForestSRC was released to CRAN and I updated PR #832 accordingly: (Hopefully) now all problems/bugs are solved, all tests are enabled (so #831 is obsolete), all randomForestSRC interfaces are up to date and we can close this monster issue...
:v:
